Scheduler run_job framework + Reload svr_conns during sched HUP + IFL…

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->

This PR consists of the following changes.

- All the run_job related code changes which ensures that scheduler sends the jobs to the right server with the target server's hostname and port. Also need to make code changes ATTR_pbs_server/ATTR_server of Job and Node objects to have the respective pbs_server conf variable value. This includes both Server/Scheduler side changes.
- Fix for intermittent IFL call failures(especially before new server comes up fully if the IFL queries go to new server)
- Made Scheduler code changes **to reload svr_conns table** with the right data. After this I am able to test these changes by modifying PBS_SERVER_INSTANCES in pbs_conf without stopping PBS and verified that now sched reloads all svr_conns table.
- Even after the above changes pbs_add_server script started failing due to couple of race conditions when Server contacting scheduler. Find the root causes for the same and fixed it.
- Handling Misconfiguration at pbs_sched (for example more number of servers getting connected than configured)
-  **Making Server connect to remote Server :** Right now default scheduler does not connect to a different remote server in another complex. This was by design. Now I had to change it for Multi-Server so that two different PBS Complexes still can use single Scheduler. May be we need to rewrite some PTL test cases around this. We can take it up when we do it for mainline.
- Fixed a race condition in Scheduler Failover.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->


#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
I have tested the code changes by setting up multiple servers to the extent possible. Following are the logs for the same.
[MultiSvrTesting_v1.txt](https://github.com/subhasisb/openpbs/files/4888871/MultiSvrTesting_v1.txt)
 



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

suresh-thelkar commented:
All the test cases from Travis and AppVeyor are passed. OpenSUSE is failing due to the know issue.

suresh-thelkar commented:
Requesting @nithinj/ @subhasisb  reviews.

nithinj commented on src/scheduler/fifo.c:
```diff
@@ -1303,14 +1304,73 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
+
+		svr_id_len = strlen(svr_conns[i]->svr_id);
+		extend[svr_id_len] = '\0';
+		strncpy(extend, svr_conns[i]->svr_id, svr_id_len);
```
Why not simply do an strdup?
extend is getting freed inside IFL?

suresh-thelkar replied:
Even strdup() is also not required. I have optimised the code. Please look into the latest commits. With this code change no need to free extend.

nithinj commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2214,7 +2217,7 @@ random_srv_conn(svr_conn_t **svr_connections)
 	ind =  rand() % get_current_servers();
 
-	if (svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)
+	if ((svr_connections[ind] != NULL) && svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)
```
just check for `svr_connections[ind]`
`if (svr_connections[ind] && svr_connections[ind]->state == SVR_CONN_STATE_CONNECTED)`

nithinj commented on src/scheduler/fifo.c:
```diff
@@ -1303,14 +1304,73 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
+				return -1;
+			}
+
+			*colon_ptr = '\0';
```
instead of duping, you can also restore the colon back

suresh-thelkar replied:
I am now restoring the colon.

suresh-thelkar commented:
@nithinj, I have addressed your comments. Please take a look.

I have also added one more commit which has code changes to fix a race condition that happens when all servers try to connect to Scheduler in parallel.

@subhasisb / @nithinj, Can you please also review this new commit as well.

suresh-thelkar commented on src/include/pbs_ifl.h:
```diff
@@ -158,6 +158,7 @@ extern "C" {
 #define ATTR_state	"job_state"
 #define ATTR_queue	"queue"
 #define ATTR_server	"server"
+#define ATTR_pbs_server	"pbs_server"
```
Yes. I have renamed it to ATTR_server and did all the corresponding changes.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -1303,14 +1304,67 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
+
+			*colon_ptr = '\0';
+
+			if (strcmp(svr_conns[i]->svr_id, svr_of_node) == 0) {
```
Yes it is in fact going to the server who owns the job. This particular code is to find out the destination server and its port. Please look at the following line of code which takes care of this. 
src_svr_index = get_svr_index(svr_of_job);

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -1303,14 +1304,67 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
+			if (svr_conns[i] == NULL)
+				continue;
+
+			colon_ptr = strstr(svr_conns[i]->svr_id, ":") ;
```
Replaced strstr with strchr.

suresh-thelkar commented:
@nithinj, Thanks for sign off.
@agrawalravi90, I have implemented all your comments.
